docs/apache-airflow/common-pitfall.rst

```diff
@@ -0,0 +1,202 @@
+that. For now, make sure that the DAG object is in the global namespace for it be recognized by Airflow.
+
+.. code-block:: python
+
+        globals ()[dag_id] = DAG(...)
+
+Refer to :ref:`how to build dynamic DAGs<faq:dynamic_dag>`.
+
+Top level Python code
+----------------------
+
+While it is not recommended to write any code outside of defining Airflow constructs, Airflow does support any
+arbitrary python code as long as it does not break the DAG file processor or prolong file processing time past the
+:ref:`config:core__dagbag_import_timeout` value.
+
+A common example is the violation of the time limit when building a dynamic DAG which usually requires querying data
+from another service like a database. At the same time, the requested service is being swamped with DAG file
+processors requests for data to process the file. These unintended interactions may cause the service to deteriorate
+and eventually cause DAG file processing to fail.
+
+Refer to :ref:`DAG writing best practices<best_practice:writing_a_dag>` for more information.
+
+Double Jinja Templating
+-----------------------
+
+It is not possible to render a Jinja template within another Jinja template. This is commonly attempted in
+``user_defined_macros``.
+
+.. code-block:: python
+
+        dag = DAG(
+            ...
+            user_defined_macros={
+                'my_custom_macro': 'day={{ ds }}'
+            }
+        )
+
+        bo = BashOperator(
+            task_id='my_task',
+            bash_command="echo {{ my_custom_macro }}",
+            dag=dag
+        )
+
+This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with the execution date 2020-01-01 00:00:00.
+
+.. code-block:: python
+
+        bo = BashOperator(
+            task_id='my_task',
+            bash_command="echo day={{ ds }}",
+            dag=dag
+        )
+
+By using the ds macros directly in the template_field, the rendered value results in "day=2020-01-01".
+
+Operators and Hooks
+^^^^^^^^^^^^^^^^^^^
+
+File templating and file extensions
+-----------------------------------
+
+TemplateNotFound errors are usually due to misalignment with user expectations when passing special values to Operator
+that trigger Jinja templating. A common occurrence is with BashOperators.
+
+Given BashOperator's ``template_fields`` includes ``bash_command`` and ``template_ext`` is a non-empty list, Airflow
+will attempt to render ``bash_command`` with the contents of a file using the parameter value as the file path if
+the parameter value ends in one of the listed file extensions.
+
+.. code-block:: python
+
+        bo = BashOperator(
+            task_id='my_script',
+            bash_command="/usr/local/airflow/include/test.sh",
+            dag=dag
+        )
+
+If you wish to directly executed a bash script, you need to add a space after the script name to prevent Airflow from
+rendering the template using Jinja.
```

Comment:
This is duplicate content. It's explained in https://github.com/apache/airflow/blob/master/docs/apache-airflow/howto/operator/bash.rst#jinja-template-not-found

You can create a ref link to `bash.rst` instead.

Reply:
Ahh! Yeah, I will add a reference to `bash.rst`.

On the same note, Should I move the the relative path interaction and the interaction between `template_fields` and `template_ext` into `faq.rst` since that issue is not limited to only the BashOperator. I can reference `faq.rst` in `bash.rst` instead.